test/test_testing.py

```diff
@@ -4002,7 +4053,41 @@ def test_send_signal(self):
 class _PebbleStorageAPIsTestMixin:
     # Override this in classes using this mixin.
     # This should be set to any non-empty path, but without a trailing /.
-    prefix = None
+    prefix: str = None  # type: ignore
+
+    # Override this in classes using this mixin.
+    client: ops.pebble.Client = None  # type: ignore
+
+    if typing.TYPE_CHECKING:
+        # Provide these in classes using this mixin, generally by mixing with a TestCase.
+        def assertEqual(  # noqa
+                self,
+                first: typing.Any,
+                second: typing.Any,
+                msg: typing.Optional[str] = None):
+            ...
+
+        _cm_type = typing.ContextManager[unittest.case._AssertRaisesContext[BaseException]]
```

Comment:
@benhoyt I'm not completely certain that this is the right type for the return value of `assertRaises` when used as a context manager. I couldn't find much stdlib documentation on typing context managers, and web searches had a bunch of dated content that didn't seem to help much.

This _does_ work, in that pyright (and whatever VS Code is using, maybe also pyright) seems ok with it, but I'm not convinced that the `cm` in `with self.assertRaises(MyException) as cm:` is of the right type. In addition, some attributes are only set in `__exit__` (I assume, I didn't read the code), so, for example, `cm.exception` only exists after the `with` is complete - which is why I've also had to do `cm.exception = cast(ExceptionType, cm.exception)` where this gets used.

If you're more familiar with this or know someone who is or somewhere good to read up on it, I'd love some pointers :)

Reply:
Hmm, no, this is well outside my area of expertise. For reference, [this is how typeshed (type hints for the stdlib) annotates that](https://github.com/python/typeshed/blob/6074a8f5441f7e6cb833409f11ed315598fa67a3/stdlib/unittest/case.pyi#L146).

However, all this is a bit messy. I tried making the mixin class also inherit from `unittest.TestCase`, but the method-resolution-order algorithm doesn't like that. Can we skirt the issue and drop this messiness by doing this?

```python
class _PebbleStorageAPIsTestMixin:
    assertEqual = unittest.TestCase.assertEqual
    assertRaises = unittest.TestCase.assertRaises
    assertIsInstance = unittest.TestCase.assertIsInstance
    assertIn = unittest.TestCase.assertIn
```

I think then we can also drop the cast calls like:

```python
        cm.exception = typing.cast(pebble.PathError, cm.exception)
```

Reply:
Nice suggestion, thanks!

I'm super happy to be rid of those cast calls, I hated those. And I'm pretty sure (even more now that they are not required) that they were only needed because I had something wrong in the return type.